### PR TITLE
Cdat web plot subsetting

### PR DESCRIPTION
**Pull request by doutriaux1**
@danlipsa this is the rest of your PR that you merged before I got a chance to push it. Please review.

### CODE REVIEW

File: Packages/cdms2/Lib/avariable.py [hunk 1]
```diff
@@ -1227,15 +1224,15 @@ def specs2slices (self, speclist, force=None):
...
                 slicelist.append (slice(key,key+1))
-            elif type(key) is types.SliceType: # x[i:j:k]
+            elif isinstance(key, slice): # x[i:j:k]
                 slicelist.append(key)
             elif key is unspecified or key is None or key == ':':
                 slicelist.append (slice(0, len(self.getAxis(i))))
             elif key is Ellipsis:
                 raise CDMSError, "Misuse of ellipsis in specification."
-            elif type(key) is types.TupleType:
+            elif sinstance(key, tuple):
```
Comment (danlipsa) [inline Packages/cdms2/Lib/avariable.py]: isinstance is misspeled here. Would flake8 catch this?

File: Packages/cdms2/Lib/axis.py [hunk 2]
```diff
@@ -22,13 +22,13 @@ class AliasList (UserList):
...
+        self.data[i] = value.lower()
     def __setslice(self, i, j, values):
-        self.data[i:j] = map(lambda x: string.lower(x), values)
+        self.data[i:j] = map(lambda x: x.lower(), values)
     def append(self, value):
-        self.data.append(string.lower(value))
+        self.data.append(value.lower())
     def extend(self, values):
-        self.data.extend(map(string.lower, values))
+        self.data.extend(map("".lower, values))
```
Comment (danlipsa) [inline Packages/cdms2/Lib/axis.py]: Does this work correctly? Would basestring.lower work?

File: Packages/cdms2/Lib/cdmsobj.py [hunk 3]
```diff
@@ -549,12 +546,12 @@ def matchone(self, pattern, attname):
         """
         if attname is None:
             for attval in self.attributes.values():
-                if type(attval) is types.StringType and pattern.match(attval) is not None:
+                if isinstance(attval,basestring) and pattern.search(attval) is not None:
```
Comment (danlipsa) [inline Packages/cdms2/Lib/cdmsobj.py]: This has a different behavior: match vs search. Not sure what the correct behavior is.

File: Packages/cdms2/Lib/cdmsobj.py [hunk 4]
```diff
@@ -549,12 +546,12 @@ def matchone(self, pattern, attname):
...
                     return 1
             return 0
         elif self.attributes.has_key(attname):
             attval = self.attributes[attname]
-            return (type(attval) is types.StringType and pattern.match(attval) is not None)
+            return isinstance(attval,basestring) and pattern.search(attval) is not None
```
Comment (danlipsa) [inline Packages/cdms2/Lib/cdmsobj.py]: Same here.

File: Packages/cdms2/Lib/database.py [hunk 5]
```diff
@@ -205,7 +203,7 @@ def __del__(self):
     
     def normalizedn(self, dn):
         explodeddn = ldap.explode_dn(dn)
-        return string.join(explodeddn,',')
+        return ','.join(explodeddn)
```
Comment (danlipsa) [inline Packages/cdms2/Lib/database.py]: This reverses the order of parameters compared with original.

File: Packages/cdms2/Lib/database.py [hunk 6]
```diff
@@ -239,7 +237,7 @@ def getObjFromDataset(self, dn):
 
         # Get the parent dataset
         explodeddn = ldap.explode_dn(dn)
-        dsetdn = string.join(explodeddn[1:],',') # Dataset node is parent of variable
+        dsetdn = ','.join(explodeddn[1:]) # Dataset node is parent of variable
```
Comment (danlipsa) [inline Packages/cdms2/Lib/database.py]: Same here.

File: Packages/cdms2/Lib/slabinterface.py [hunk 7]
```diff
@@ -143,7 +143,7 @@ def showdim(self):
         for nd in range(self.rank()):
             result.append('** Dimension ' + str(nd+1) + ' **')
             result = result + self.getAxis(nd).listall(1)
-        print string.join(result, '\n')
+        print '\n'.join(result)
```
Comment (danlipsa) [inline Packages/cdms2/Lib/slabinterface.py]: Reverses order.

File: Packages/cdms2/Lib/slabinterface.py [hunk 8]
```diff
@@ -179,7 +179,7 @@ def listall (self, all=None):
     def info(self, flag=None, device=None):
         "Write info about slab; include dimension values and weights if flag"
         if device is None: device = sys.stdout
-        device.write(string.join(self.listall(all=flag), "\n"))
+        device.write('\n'.join(self.listall(all=flag)))
```
Comment (danlipsa) [inline Packages/cdms2/Lib/slabinterface.py]: reverses order

File: Packages/cdms2/Script/cdscan [hunk 9]
```diff
@@ -426,7 +424,7 @@ def disambig(name, dict, num, comparator, value):
         uniqname = '%s_%d'%(name,num)
         if dict.has_key(uniqname) and comparator(value, dict[uniqname]):
             trial_name = uniqname
-            for letter in string.lowercase:
+            for letter in 'abcdefghijklmnopqrstuvwxyz':
```
Comment (danlipsa) [inline Packages/cdms2/Script/cdscan]: Use letter.islower()

Comment (danlipsa) [inline Packages/cdms2/Script/cdscan]: There is a string.ascii_lowercase

File: Packages/cdms2/Lib/avariable.py [hunk 10]
```diff
@@ -1523,8 +1514,8 @@ def orderparse (order):
         elif r == '...':
             r = Ellipsis
         elif len(r) == 1:
-            if r in string.digits:
-                r = string.atoi(r)
+            if r in '0123456789':
```
Comment (danlipsa) [inline Packages/cdms2/Lib/avariable.py]: Use r.isdigit()

**Comment by danlipsa** [comment]
Nice cleanup! See my comments. @aashish24 Please review.

File: Packages/cdms2/Lib/cdmsobj.py (hunk 3 above)
Comment (doutriaux1) [inline Packages/cdms2/Lib/cdmsobj.py]: oops copy/pasting too much! Nice catch!

File: Packages/cdms2/Lib/avariable.py (hunk 1 above)
Comment (doutriaux1) [inline Packages/cdms2/Lib/avariable.py]: yep i need to run flake8 on this.

File: Packages/cdms2/Lib/database.py (hunk 5 above)
Comment (doutriaux1) [inline Packages/cdms2/Lib/database.py]: i dont think so:
```python
>>> string.join(["A","B"],"-")
'A-B'
>>> "-".join(["A","B"])
'A-B'
```

File: Packages/cdms2/Lib/database.py (hunk 6 above)
Comment (doutriaux1) [inline Packages/cdms2/Lib/database.py]: same here :wink:

File: Packages/cdms2/Lib/slabinterface.py (hunk 7 above)
Comment (doutriaux1) [inline Packages/cdms2/Lib/slabinterface.py]: nope

File: Packages/cdms2/Lib/slabinterface.py (hunk 8 above)
Comment (doutriaux1) [inline Packages/cdms2/Lib/slabinterface.py]: still nope

File: Packages/cdms2/Script/cdscan (hunk 9 above)
Comment (doutriaux1) [inline Packages/cdms2/Script/cdscan]: yes but I do not want to use the string module

**Comment by danlipsa** [comment]
@doutriaux1 What about the comments where I said it reverses the order. I don't think that happens actually, but rather it uses the default separator: ' ' rather than \n or ,

**Comment by doutriaux1** [comment]
@danlipsa I commented on this, it doesn't reverse the order, look at my sample code in my comments.
```python
In [1]: import string

In [2]: string.join(["A","B"],"-")
Out[2]: 'A-B'

In [3]: "-".join(["A","B"])
Out[3]: 'A-B'
```

**Comment by danlipsa** [comment]
@doutriaux1 I see. Confusing syntax for the join in statement 3! You would expect a member function of an object to be applied to the object, so the result would be - A B. I like the syntax in 2 a lot better.

**Comment by danlipsa** [comment]
@doutriaux1 LGTM. :+1:

**Comment by aashish24** [comment]
@doutriaux1 looks like we need to update the branch to resolve the conflict.

**Comment by aashish24** [comment]
@doutriaux1 @danlipsa should we update this branch and merge it?

**Comment by aashish24** [comment]
@doutriaux1 @danlipsa same here. I believe we all did LGTM. Any hold on this branch?

**Pull request merged**